### PR TITLE
[RFR] Migrate ra-core controllers to TypeScript

### PR DESCRIPTION
**Pull request by djhi**
- [x] Fields
- [x] Inputs
- [x] Views

### CODE REVIEW

File: packages/ra-core/src/actions/accumulateActions.ts [hunk 1]
```diff
@@ -6,16 +6,19 @@ export interface CrudGetManyAccumulateAction {
...
-        ids: [];
+        ids: any[];
     };
     readonly meta: {
         accumulate: any;
     };
 }
 
+// Used to type the dispatcher function (the one injected by mapDispatchToProps)
+export type CrudGetManyAccumulate = (resource: string, ids: any[]) => void;
```
Comment (djhi) [inline packages/ra-core/src/actions/accumulateActions.ts]: Maybe this should be named `DispatchCrudGetManyAccumulate`. @fzaninotto, what do you think?

Comment (fzaninotto) [reply packages/ra-core/src/actions/accumulateActions.ts]: "Dispatch" should only be used when the action creator is decorated by redux. Otherwise, we may end up with a `dispatchDispatchCrudGetManyAccumulate` prop name.

Comment (djhi) [reply packages/ra-core/src/actions/accumulateActions.ts]: Well, this is actually the type used for the function passed as prop by the `mapToDispatch` argument of connect

Comment (fzaninotto) [reply packages/ra-core/src/actions/accumulateActions.ts]: understood, but it's not a dispatcher at that step

Comment (djhi) [reply packages/ra-core/src/actions/accumulateActions.ts]: Hu ? It is

File: packages/ra-core/src/actions/accumulateActions.ts [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
+import { Pagination, Sort } from '../types';
 
 export const CRUD_GET_MANY_ACCUMULATE = 'RA/CRUD_GET_MANY_ACCUMULATE';
 
 export interface CrudGetManyAccumulateAction {
     readonly type: typeof CRUD_GET_MANY_ACCUMULATE;
     readonly payload: {
         resource: string;
-        ids: [];
+        ids: any[];
```
Comment (fzaninotto) [inline packages/ra-core/src/actions/accumulateActions.ts]: why not `Identifier[]`?

File: packages/ra-core/src/actions/accumulateActions.ts [hunk 3]
```diff
@@ -15,7 +16,7 @@ export interface CrudGetManyAccumulateAction {
 
 export const crudGetManyAccumulate = (
     resource: string,
-    ids: []
+    ids: any[]
```
Comment (fzaninotto) [inline packages/ra-core/src/actions/accumulateActions.ts]: Same remark

File: packages/ra-core/src/actions/accumulateActions.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/actions/accumulateActions.ts]: Good point, it should be `Identifier[]`

File: packages/ra-core/src/controller/CreateController.tsx [hunk 4]
```diff
@@ -50,7 +81,13 @@ import checkMinimumRequiredProps from './checkMinimumRequiredProps';
  *     );
  *     export default App;
  */
-export class CreateController extends Component {
+export class CreateControllerView extends Component<Props> {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/CreateController.tsx]: I don't agree with the renaming. In that particular case, it is not a view, it is a controller. If you're looking for a better name, try `CreateControllerCore` or `UnconnectedCreateController`.

Same for other renamings in ra-core controllers.

File: packages/ra-core/src/controller/CreateController.tsx [hunk 5]
```diff
@@ -50,7 +81,13 @@ import checkMinimumRequiredProps from './checkMinimumRequiredProps';
...
+    public static defaultProps = {
+        record: {},
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/CreateController.tsx]: `Partial<Record>`?

File: packages/ra-core/src/controller/CreateController.tsx [hunk 6]
```diff
@@ -1,13 +1,44 @@
...
+    hasCreate: boolean;
+    hasEdit: boolean;
+    hasList: boolean;
+    hasShow: boolean;
+    isLoading: boolean;
+    location: Location;
+    match: Match;
+    record?: Record;
+    resource: string;
+    title: string | ReactNode;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/CreateController.tsx]: You didn't transpose the propTypes correctly. `title`, for instance, was not required. Please double check for all components.

File: packages/ra-core/src/controller/EditController.tsx [hunk 7]
```diff
@@ -1,13 +1,53 @@
...
+    crudGetOne,
+    crudUpdate,
+    startUndoable as startUndoableAction,
+} from '../actions';
 import { REDUX_FORM_NAME } from '../form';
 import checkMinimumRequiredProps from './checkMinimumRequiredProps';
+import { Translate, Record, Dispatch } from '../types';
+import { RedirectionSideEffect } from '../sideEffect';
+
+interface ChildrenFuncParams {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/EditController.tsx]: :+1: This is making the code much clearer

File: packages/ra-core/src/controller/ListController.tsx [hunk 8]
```diff
@@ -70,9 +145,29 @@ import checkMinimumRequiredProps from './checkMinimumRequiredProps';
...
+        perPage: 10,
+        sort: {
+            field: 'id',
+            order: SORT_DESC,
+        },
+    };
+
     state = {};
 
+    setFilters = debounce(filters => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/ListController.tsx]: why did you move that function?

File: packages/ra-core/src/controller/ListController.tsx [hunk 9]
```diff
@@ -310,9 +395,11 @@ export class ListController extends Component {
...
             onToggleItem: this.handleToggleItem,
             onUnselectItems: this.handleUnselectItems,
-            page: parseInt(query.page || 1, 10),
-            perPage: parseInt(query.perPage, 10),
-            refresh: this.refresh,
+            page:
+                (typeof query.page === 'string'
+                    ? parseInt(query.page, 10)
+                    : query.page) || 1,
+            perPage: parseInt(query.perPage.toString(), 10),
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/ListController.tsx]: please use the same test as in the previous line. I prefer not casting to string to get a number in the end

File: packages/ra-core/src/controller/ShowController.tsx [hunk 10]
```diff
@@ -1,11 +1,41 @@
...
+    children: (params: ChildrenFuncParams) => ReactNode;
+    crudGetOne: Dispatch<typeof crudGetOneAction>;
+    record?: Record;
+    hasCreate: boolean;
+    hasEdit: boolean;
+    hasShow: boolean;
+    hasList: boolean;
+    id: string;
+    isLoading: boolean;
+    location: object;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/ShowController.tsx]: I think we can use the React Router types here

File: packages/ra-core/src/controller/field/ReferenceArrayFieldController.tsx [hunk 11]
```diff
@@ -1,10 +1,30 @@
...
 import { connect } from 'react-redux';
 import get from 'lodash/get';
 
 import { crudGetManyAccumulate as crudGetManyAccumulateAction } from '../../actions';
 import { getReferencesByIds } from '../../reducer/admin/references/oneToMany';
+import { ReduxState, Record, RecordMap, Dispatch } from '../../types';
+
+interface ChildrenFuncParams {
+    loadedOnce: boolean;
+    ids: any[];
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/field/ReferenceArrayFieldController.tsx]: identifiers

File: packages/ra-core/src/controller/field/ReferenceArrayFieldController.tsx [hunk 12]
```diff
@@ -1,10 +1,30 @@
...
+    data: RecordMap;
+    referenceBasePath: string;
+    currentSort: any;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/field/ReferenceArrayFieldController.tsx]: Sort

File: packages/ra-core/src/controller/field/ReferenceManyFieldController.tsx [hunk 13]
```diff
@@ -61,11 +93,19 @@ import {
...
-    }
+export class ReferenceManyFieldControllerView extends Component<Props> {
+    public static defaultProps: Partial<Props> = {
+        filter: {},
+        perPage: 25,
+        sort: { field: 'id', order: 'DESC' },
+        source: 'id',
+    };
+
+    public state = {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/field/ReferenceManyFieldController.tsx]: Could you type it?

File: packages/ra-core/src/controller/input/ReferenceInputController.tsx [hunk 14]
```diff
@@ -16,9 +15,56 @@ import {
...
+
+interface Props {
+    allowEmpty?: boolean;
+    basePath: string;
+    children: (params: ChildrenFuncParams) => ReactNode;
+    crudGetMatchingAccumulate: Dispatch<typeof crudGetMatchingAccumulateAction>;
+    crudGetManyAccumulate: Dispatch<typeof crudGetManyAccumulateAction>;
+    filter?: object;
+    filterToQuery: (filter: {}) => any;
+    input?: {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceInputController.tsx]: don't we have a type from redux-form for that?

File: packages/ra-core/src/reducer/admin/references/oneToMany.ts [hunk 15]
```diff
@@ -75,7 +75,11 @@ export const getReferences = (state: ReduxState, reference, relatedTo) => {
         }, {});
 };
 
-export const getReferencesByIds = (state: ReduxState, reference, ids) => {
+export const getReferencesByIds = (
+    state: ReduxState,
+    reference: string,
+    ids: any[]
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/admin/references/oneToMany.ts]: identifier

File: packages/ra-core/src/reducer/admin/resource/list/params.ts [hunk 16]
```diff
@@ -30,7 +30,7 @@ const paramsReducer: Reducer<State> = (
 ) => {
     switch (action.type) {
         case CRUD_CHANGE_LIST_PARAMS:
-            return action.payload;
+            return { ...previousState, ...action.payload };
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/admin/resource/list/params.ts]: please revert

File: packages/ra-core/src/types.ts [hunk 17]
```diff
@@ -4,6 +4,10 @@ export interface Record {
     [key: string]: any;
 }
 
+export interface RecordMap {
+    [id: string]: Record;
```
Comment (fzaninotto) [inline packages/ra-core/src/types.ts]: unfortunately, there is a bug in TypeScript that doesn't let you use Identifer as a key type, but I'm afraid this will break if the id is a number

File: packages/ra-core/src/controller/CreateController.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-core/src/controller/CreateController.tsx]: Yes and actually, that's a prop of the `ra-ui-materialui` `Create` component.

File: packages/ra-core/src/controller/ListController.tsx (hunk 8 above)
Comment (djhi) [reply packages/ra-core/src/controller/ListController.tsx]: That's not me, tslint did it. I suppose it's because it considers it a private field and not a private function

File: packages/ra-core/src/controller/input/ReferenceInputController.tsx (hunk 14 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceInputController.tsx]: We would have to include the redux-form types in our dev deps and last time I tried that, they were not good

File: packages/ra-core/src/controller/CreateController.tsx (hunk 5 above)
Comment (djhi) [reply packages/ra-core/src/controller/CreateController.tsx]: I type `defaultProps` instead

File: packages/ra-core/src/controller/input/ReferenceInputController.tsx (hunk 14 above)
Comment (djhi) [reply packages/ra-core/src/controller/input/ReferenceInputController.tsx]: I added them and it seems ok

File: packages/ra-core/src/controller/CreateController.tsx (hunk 6 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/CreateController.tsx]: I saw you remove the title prop here, but I suspect you didn't cross check the isRequired attribute on all props. hasEdit, for instance, is required in your version but was not require din propTypes

File: packages/ra-core/src/controller/ShowController.tsx [hunk 18]
```diff
@@ -1,11 +1,41 @@
...
+    translate: Translate;
+    version: number;
+}
+
+interface Props {
+    basePath: string;
+    children: (params: ChildrenFuncParams) => ReactNode;
+    crudGetOne: Dispatch<typeof crudGetOneAction>;
+    record?: Record;
+    hasCreate: boolean;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/ShowController.tsx]: still missing optional here

**Comment by fzaninotto** [comment]
\o/

**Comment by djhi** [comment]
Wait before we test them. I'm sure we're going to be spanked hard by TS when using those definitions in the other packages :cold_sweat:
